GPU/Common/TextureCacheCommon.cpp

```diff
@@ -1439,145 +1562,175 @@ void TextureCacheCommon::DecodeTextureLevel(u8 *out, int outPitch, GETextureForm
 				if (expandTo32bit && !reverseColors) {
 					// We simply expand the CLUT to 32-bit, then we deindex as usual. Probably the fastest way.
 					ConvertFormatToRGBA8888(clutformat, expandClut_, clut, 16);
+					fullAlphaMask = 0xFF000000;
 					for (int y = 0; y < h; ++y) {
-						DeIndexTexture4((u32 *)(out + outPitch * y), texptr + (bufw * y) / 2, w, expandClut_);
+						DeIndexTexture4<u32>((u32 *)(out + outPitch * y), texptr + (bufw * y) / 2, w, expandClut_, &alphaSum);
 					}
 				} else {
+					fullAlphaMask = ClutFormatToFullAlpha(clutformat);
 					for (int y = 0; y < h; ++y) {
-						DeIndexTexture4((u16 *)(out + outPitch * y), texptr + (bufw * y) / 2, w, clut);
+						DeIndexTexture4<u16>((u16 *)(out + outPitch * y), texptr + (bufw * y) / 2, w, clut, &alphaSum);
 					}
 				}
 			}
+
+			if (clutformat == GE_CMODE_16BIT_BGR5650) {
+				// Our formula at the end of the function can't handle this cast so we return early.
+				return CHECKALPHA_FULL;
+			}
 		}
 		break;
 
 		case GE_CMODE_32BIT_ABGR8888:
 		{
 			const u32 *clut = GetCurrentClut<u32>() + clutSharingOffset;
+			fullAlphaMask = 0xFF000000;
 			for (int y = 0; y < h; ++y) {
-				DeIndexTexture4((u32 *)(out + outPitch * y), texptr + (bufw * y) / 2, w, clut);
+				DeIndexTexture4<u32>((u32 *)(out + outPitch * y), texptr + (bufw * y) / 2, w, clut, &alphaSum);
 			}
 		}
 		break;
 
 		default:
 			ERROR_LOG_REPORT(G3D, "Unknown CLUT4 texture mode %d", gstate.getClutPaletteFormat());
-			return;
+			return CHECKALPHA_ANY;
 		}
 	}
 	break;
 
 	case GE_TFMT_CLUT8:
-		ReadIndexedTex(out, outPitch, level, texptr, 1, bufw, expandTo32bit);
-		break;
+		return ReadIndexedTex(out, outPitch, level, texptr, 1, bufw, expandTo32bit);
 
 	case GE_TFMT_CLUT16:
-		ReadIndexedTex(out, outPitch, level, texptr, 2, bufw, expandTo32bit);
-		break;
+		return ReadIndexedTex(out, outPitch, level, texptr, 2, bufw, expandTo32bit);
 
 	case GE_TFMT_CLUT32:
-		ReadIndexedTex(out, outPitch, level, texptr, 4, bufw, expandTo32bit);
-		break;
+		return ReadIndexedTex(out, outPitch, level, texptr, 4, bufw, expandTo32bit);
 
 	case GE_TFMT_4444:
 	case GE_TFMT_5551:
 	case GE_TFMT_5650:
 		if (!swizzled) {
 			// Just a simple copy, we swizzle the color format.
 			if (reverseColors) {
+				// Just check the input's alpha to reuse code. TODO: make a specialized ReverseColors that checks as we go.
+				fullAlphaMask = TfmtRawToFullAlpha(format);
```

Comment:
Shouldn't this be above `if (reverseColors)`?  It's currently missing in the `expandTo32bit` case.

-[Unknown]

Reply:
Yup, fixing